econml/sklearn_extensions.py

```diff
@@ -0,0 +1,975 @@
+        y : array, shape (n_samples,)
+            Target. Will be cast to X's dtype if necessary
+
+        sample_weight : numpy array of shape [n_samples]
+                        Individual weights for each sample.
+                        The weights will be normalized internally.
+
+        check_input : boolean, (default=True)
+            Allow to bypass several input checking.
+            Don't use this parameter unless you know what you do.
+        """
+        alpha_grid = [0.01, 0.02, 0.03, 0.06, 0.1, 0.2, 0.3, 0.5, 0.8, 1]
+        self.selected_alpha_ = None
+        if self.alpha == 'auto':
+            # Select optimal penalty
+            self.alpha = self._get_optimal_alpha(
+                alpha_grid, X, y, sample_weight)
+            self.selected_alpha_ = self.alpha
+        else:
+            # Warn about consistency
+            warnings.warn("Setting a suboptimal alpha can lead to miscalibrated confidence intervals. "
+                          "We recommend setting alpha='auto' for optimality.")
+
+        # Convert X, y into numpy arrays
+        X, y = check_X_y(X, y, y_numeric=True, multi_output=False)
+        # Fit weighted lasso with user input
+        super(DebiasedLasso, self).fit(X, y, sample_weight, check_input)
+        # Center X, y
+        X, y, X_offset, y_offset, X_scale = self._preprocess_data(
+            X, y, fit_intercept=self.fit_intercept, normalize=False,
+            copy=self.copy_X, check_input=check_input, sample_weight=sample_weight, return_mean=True)
+
+        # Calculate quantities that will be used later on. Account for centered data
+        y_pred = self.predict(X) - self.intercept_
+        self._theta_hat = self._get_theta_hat(X, sample_weight)
+        self._X_offset = X_offset
+
+        # Calculate coefficient and error variance
+        num_nonzero_coefs = np.sum(self.coef_ != 0)
+        self._error_variance = np.average((y - y_pred)**2, weights=sample_weight) / \
+            (1 - num_nonzero_coefs / X.shape[0])
+        self._mean_error_variance = self._error_variance / X.shape[0]
+        self._coef_variance = self._get_unscaled_coef_var(
+            X, self._theta_hat, sample_weight) * self._error_variance
+
+        # Add coefficient correction
+        coef_correction = self._get_coef_correction(
+            X, y, y_pred, sample_weight, self._theta_hat)
+        self.coef_ += coef_correction
+
+        # Set coefficients and intercept standard errors
+        self.coef_std_err_ = np.sqrt(np.diag(self._coef_variance))
+        if self.fit_intercept:
+            self.intercept_std_err_ = np.sqrt(
+                np.matmul(np.matmul(self._X_offset, self._coef_variance), self._X_offset) +
+                self._mean_error_variance
+            )
+        else:
+            self.intercept_std_err_ = 0
+
+        # Set intercept
+        self._set_intercept(X_offset, y_offset, X_scale)
+        # Return alpha to 'auto' state
+        if self.selected_alpha_ is not None:
```

Comment:
Is there a risk that an exception will be thrown in the interim, leaving the state incorrect for a subsequent call to fit?  If so, might be better to put this within an `except` block.